astropy/stats/spatial.py

```diff
@@ -0,0 +1,314 @@
+       <https://en.wikipedia.org/wiki/Spatial_descriptive_statistics>
+    .. [2] Package spatstat.
+       <https://cran.r-project.org/web/packages/spatstat/spatstat.pdf>
+    .. [3] Cressie, N.A.C. (1991). Statistics for Spatial Data,
+       Wiley, New York.
+    .. [4] Stoyan, D., Stoyan, H. (1992). Fractals, Random Shapes and
+       Point Fields, Akademie Verlag GmbH, Chichester.
+    """
+
+    def __init__(self, area, x_max=None, y_max=None, x_min=0, y_min=0):
+        self.area = area
+        self.x_max = x_max
+        self.y_max = y_max
+        self.x_min = x_min
+        self.y_min = y_min
+
+    @property
+    def area(self):
+        return self._area
+
+    @area.setter
+    def area(self, value):
+        if isinstance(value, (float, int)) and value > 0:
+            self._area = value
+        else:
+            raise ValueError('area is expected to be a positive number. '
+                             'Got {}.'.format(value))
+
+    @property
+    def y_max(self):
+        return self._y_max
+
+    @y_max.setter
+    def y_max(self, value):
+        if value is None or (isinstance(value, (float, int)) and value > 0):
+            self._y_max = value
+        else:
+            raise ValueError('y_max is expected to be a positive number '
+                             'or None. Got {}.'.format(value))
+
+    @property
+    def x_max(self):
+        return self._x_max
+
+    @x_max.setter
+    def x_max(self, value):
+        if value is None or (isinstance(value, (float, int)) and value > 0):
+            self._x_max = value
+        else:
+            raise ValueError('x_max is expected to be a positive number '
+                             'or None. Got {}.'.format(value))
+
+    @property
+    def y_min(self):
+        return self._y_min
+
+    @y_min.setter
+    def y_min(self, value):
+        if isinstance(value, (float, int)) and value >= 0:
+            self._y_min = value
+        else:
+            raise ValueError('y_min is expected to be a nonnegative number. '
+                             'Got {}.'.format(value))
+
+    @property
+    def x_min(self):
+        return self._x_min
+
+    @x_min.setter
+    def x_min(self, value):
+        if isinstance(value, (float, int)) and value >= 0:
+            self._x_min = value
+        else:
+            raise ValueError('x_min is expected to be a nonnegative number. '
+                             'Got {}.'.format(value))
+
+    def __call__(self, data, radii, mode='none'):
+        return self.evaluate(data=data, radii=radii, mode=mode)
+
+    def _pairwise_diffs(self, data):
+        npts = len(data)
+        diff = np.zeros(shape=(npts * (npts - 1) // 2, 2), dtype=np.double)
+        k = 0
+        for i in range(npts - 1):
+            for j in range(i + 1, npts):
+                diff[k] = abs(data[i] - data[j])
+                k = k + 1
+
+        return diff
+
+    def _pairwise_distances(self, data):
+        npts = len(data)
```

Comment:
`scipy` contains some distance calculation routines, maybe one could resort to them if `scipy` is installed:

```
return scipy.spatial.distance.pdist(data)
```